### PR TITLE
Improve paredit in REPL window

### LINKED ISSUE #255: Overly aggressive paredit in REPL window
If I type `[1 2 3]` in the REPL window and then decide that I want it to be `(shuffle [1 2 3])` then if I click at the beginning and type `(` then it adds `)` right after it automatically, which is the sort of thing I've come to expect both in Calva and in other environments... *but* unlike when this happens in other contexts there appears to be no way to delete the `)` no matter what I do.

I know that I'm an outlier in not personally liking any automatic typing of closing brackets ever, and also in not wanting to learn/use Paredit commands (I prefer ordinary typing), and I don't expect the behavior of Calva to change in response to my preferences! But the middle ground adopted in Calva editor windows -- adding closing brackets but letting me delete them -- works well enough for me, and presumably also for others.

So my request is that the REPL window acts like the editor windows in this respect.

### PR DESCRIPTION
**Pull request by cfehse**
## What has Changed?

- Edit message/event mechanism to populate the paredit keymap setting into the REPL webview.
- Made the Paredit implemention in the webview obey the keymap setting and behave like the editor according to this setting.
- Change the completion overlay in the webview to allow the selection of an item using the `tab` or `return` key and close the overlay after the selection (behavior like in the vscode editor)
- Changed the statusbar item for the paredit to toggle between the `strict` and `original` keymap. If the keymap is set to `none` the item is disabled.

Fixes #255, #394 

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Made sure I am directing this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I am changed the default PR base branch, so that it is not `master`. (Sorry for the nagging.)
- [x] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.) You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. (For now you'll need to opt in to the CircleCI _New Experience_ UI to see the Artifacts tab, because bug.)
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.

## The Calva TEAM PR Checklist:

Before merging we (at least one of us) have:

- [x] Made sure the PR is directed at the `dev` branch (unless reasons).
- [x] Read the source changes.
- [x] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [x] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] If need be, had a chat within the team about particular changes.

Ping @pez, @kstehn, @cfehse

### CODE REVIEW

**Comment by cfehse** [comment]
@PEZ 

I have tried to restrict the commands in the REPL implementation of Paredit (`src/webview/repl-console.ts`). I don't know if I did it right. Please take a close look at all the changes.

Thanks!

**Comment by PEZ** [comment]
Sweet! Will have a look at this.

**Comment by PEZ** [comment]
It works like a charm! Both the paredit mode and the enter in completion overlay. That latter thing is saving my sanity, I notice. I so often forget that it doesn't work to use enter, and then strange things happens instead.

I'm pleased that you choose to use the toggle button for strict mode. :heart: Speaking of that button... :smile: : 

1. When the keymap is set to `none`, the button doesn't work. Maybe that is ok. But it says that Paredit is disabled, which is wrong. Maybe it should just say **Paredit keymap is set to `none`, toggling strict mode is disabled**
1. The button should show what mode it is using. So maybe it can be w/o parens in `original` mode and with in `strict`.
1. I want to be able to toggle this with a command, like we have with the `pprint` toggle. Maybe it should be called **Toggle Paredit Mode to `<the-other-mode>`** (Is it even possible to change the text? otherwise, just skip the ”to” part), default shortcut could be `ctrl+alt+c -`
1. I'd like to have the button to the left, together with the other Calva buttons.
    * I'd also like to have the `pprint` button there, but maybe that's OT for this PR :smile:

Will continue testing and try find time to read the source changes too.

**Comment by cfehse** [comment]
>When the keymap is set to none, the button doesn't work. Maybe that is ok. But it says that Paredit is disabled, which is wrong. Maybe it should just say Paredit keymap is set to none, toggling strict mode is disabled

Done.

>The button should show what mode it is using. So maybe it can be w/o parens in original mode and with in strict.

Done.

>I want to be able to toggle this with a command, like we have with the pprint toggle. Maybe it should be called Toggle Paredit Mode to <the-other-mode> (Is it even possible to change the text? otherwise, just skip the ”to” part), default shortcut could be ctrl+alt+c -

Done without changing the title - which is not possible in the moment, I think.

>I'd like to have the button to the left, together with the other Calva buttons

Done for both buttons.

File: src/webview/repl-console.ts [hunk 1]
```diff
@@ -84,6 +92,34 @@ export class ReplConsole {
         }
     }
 
+    private isKeyMay(values: ReplPareditKeyMap[]) :boolean {
```
Comment (PEZ) [inline src/webview/repl-console.ts]: Typo, should be: `isKeyMap`, right?

File: src/webview/repl-console.ts [hunk 2]
```diff
@@ -323,16 +359,20 @@ export class ReplConsole {
 
     commands = {
         "raise-sexp": () => {
-            this.readline.withUndo(() => {
-                paredit.raiseSexp(this.readline);
-                this.readline.repaint();
-            });
+            if (this.isKeyMay([ReplPareditKeyMap.ORIGINAL, ReplPareditKeyMap.STRICT])) {
```
Comment (PEZ) [inline src/webview/repl-console.ts]: Why do we check this here? And, why only on some commands and not on others?

File: src/webview/repl-console.ts [hunk 3]
```diff
@@ -481,10 +539,15 @@ export class ReplConsole {
...
-            this.readline.withUndo(() => {
-                paredit.backspace(this.readline);
+            if (this.isKeyMay([ReplPareditKeyMap.STRICT])) {
+                this.readline.withUndo(() => {
+                    paredit.backspace(this.readline);
+                    this.readline.repaint()
+                })
+            } else {
+                this.readline.backspace();
                 this.readline.repaint()
```
Comment (PEZ) [inline src/webview/repl-console.ts]: How come we do not do this `withUndo`? (I tested it and I can undo also for normal mode, so this is probably correct. I am just wondering.)

File: src/repl-window.ts [hunk 4]
```diff
@@ -93,9 +97,13 @@ class REPLWindow {
...
         })
 
+        paredit.onPareditKeyMapChanged((keymap) => {
+            this.postMessage({ type: "paredit-keymap", keymap: keymap });
+        })
+
         this.panel.onDidDispose((e) => {
             if (this.evaluation)
-                this.evaluation.interrupt();
+                this.evaluation.interrupt().catch(() => {});
```
Comment (PEZ) [inline src/repl-window.ts]: This is a reminder to us, as a team: Lots of the repl window code has these `bare` if statements. When we change stuff in those we should put the braces in, b/c that is part of the, (yet very secret)  code convention we have.

**Comment by PEZ** [comment]
I think the code changes are good. (Put some comments here and there for details, but as a whole it makes sense.) Will test the latest changes and get back with my findings from that.

File: src/webview/repl-console.ts (hunk 2 above)
Comment (cfehse) [reply src/webview/repl-console.ts]: I tried to mimic the when clauses in the package.json. I put the check around all commands where paredit is involved. I have no idea if this makes sense.

File: src/webview/repl-console.ts (hunk 1 above)
Comment (cfehse) [reply src/webview/repl-console.ts]: Yes. *gg*

File: src/webview/repl-console.ts (hunk 3 above)
Comment (cfehse) [reply src/webview/repl-console.ts]: It comes because I forget to do so. No idea why it works. *gg*

File: src/repl-window.ts (hunk 4 above)
Comment (cfehse) [reply src/repl-window.ts]: Good point. That specific change was to remove the linter warning.

File: src/webview/repl-console.ts (hunk 2 above)
Comment (PEZ) [reply src/webview/repl-console.ts]: I don't think it makes sense. The when clause is in effect also in the REPL window. Try it in 2.0.49, setting the keymap to `none` disables Paredit shortcuts in the REPL window.

Comment (cfehse) [reply src/webview/repl-console.ts]: Okay what should I do then? Remove the check everywhere except backspace and delete?

Comment (cfehse) [reply src/webview/repl-console.ts]: Or only at 'raise-sexp'? Why do the shortcuts from package.json work in the repl?

**Comment by cfehse** [comment]
@PEZ 

I made the suggested changes from our discussion yesterday.

**Pull request merged**